File: python/pysearch/searchoptions.py
# -*- coding: utf-8 -*-
################################################################################
#
# searchoptions.py
#
# class SearchOptions: defines the available command-line options and
#                      corresponding utility methods
#
################################################################################
from collections import deque
from io import StringIO
import json
import os
import sys
import xml.dom.minidom as minidom

from common import get_text
from config import SEARCHOPTIONSPATH
from searchexception import SearchException
from searchoption import SearchOption
from searchsettings import SearchSettings


class SearchOptions(object):
    """class to provide usage info and parse command-line arguments into settings"""

    def __init__(self):
        self.options = []
        self.set_dicts()
        self.set_options_from_xml()

    def set_dicts(self):
        self.arg_action_dict = {
            'in-archiveext':
                lambda x, settings:
                    settings.add_exts(x, 'in_archiveextensions'),
            'in-archivefilepattern':
                lambda x, settings:
                    settings.add_patterns(x, 'in_archivefilepatterns'),
            'in-dirpattern':
                lambda x, settings:
                    settings.add_patterns(x, 'in_dirpatterns'),
            'in-ext':
                lambda x, settings:
                    settings.add_exts(x, 'in_extensions'),
            'in-filepattern':
                lambda x, settings:
                    settings.add_patterns(x, 'in_filepatterns'),
            'in-filetype':
                lambda x, settings:
                    settings.add_filetypes(x, 'in_filetypes'),
            'in-linesafterpattern':
                lambda x, settings:
                    settings.add_patterns(x, 'in_linesafterpatterns'),
            'in-linesbeforepattern':
                lambda x, settings:
                    settings.add_patterns(x, 'in_linesbeforepatterns'),
            'linesafter':
                lambda x, settings:
                    settings.set_property('linesafter', int(x)),
            'linesaftertopattern':
                lambda x, settings:
                    settings.add_patterns(x, 'linesaftertopatterns'),
            'linesafteruntilpattern':
                lambda x, settings:
                    settings.add_patterns(x, 'linesafteruntilpatterns'),
            'linesbefore':
                lambda x, settings:
                    settings.set_property('linesbefore', int(x)),
            'maxlinelength':
                lambda x, settings:
                    settings.set_property('maxlinelength', int(x)),
            'out-archiveext':
                lambda x, settings:
                    settings.add_exts(x, 'out_archiveextensions'),
            'out-archivefilepattern':
                lambda x, settings:
                    settings.add_patterns(x, 'out_archivefilepatterns'),
            'out-dirpattern':
                lambda x, settings:
                    settings.add_patterns(x, 'out_dirpatterns'),
            'out-ext':
                lambda x, settings:
                    settings.add_exts(x, 'out_extensions'),
            'out-filepattern':
                lambda x, settings:
                    settings.add_patterns(x, 'out_filepatterns'),
            'out-filetype':
                lambda x, settings:
                    settings.add_filetypes(x, 'out_filetypes'),
            'out-linesafterpattern':
                lambda x, settings:
                    settings.add_patterns(x, 'out_linesafterpatterns'),
            'out-linesbeforepattern':
                lambda x, settings:
                    settings.add_patterns(x, 'out_linesbeforepatterns'),
            'search':
                lambda x, settings:
                    settings.add_patterns(x, 'searchpatterns'),
            'settings-file':
                lambda x, settings:
                    self.settings_from_file(x, settings)
        }

        self.bool_flag_action_dict = {
            'allmatches':
                lambda b, settings:
                    settings.set_property('firstmatch', not b),
            'archivesonly':
                lambda b, settings:
                    settings.set_property('archivesonly', b),
            'debug':
                lambda b, settings:
                    settings.set_property('debug', b),
            'excludehidden':
                lambda b, settings:
                    settings.set_property('excludehidden', b),
            'firstmatch':
                lambda b, settings:
                    settings.set_property('firstmatch', b),
            'help':
                lambda b, settings:
                    settings.set_property('printusage', b),
            'includehidden':
                lambda b, settings:
                    settings.set_property('excludehidden', not b),
            'listdirs':
                lambda b, settings:
                    settings.set_property('listdirs', b),
            'listfiles':
                lambda b, settings:
                    settings.set_property('listfiles', b),
            'listlines':
                lambda b, settings:
                    settings.set_property('listlines', b),
            'multilinesearch':
                lambda b, settings:
                    settings.set_property('multilinesearch', b),
            'noprintmatches':
                lambda b, settings:
                    settings.set_property('printresults', not b),
            'norecursive':
                lambda b, settings:
                    settings.set_property('recursive', not b),
            'nosearcharchives':
                lambda b, settings:
                    settings.set_property('searcharchives', not b),
            'printmatches':
                lambda b, settings:
                    settings.set_property('printresults', b),
            'recursive':
                lambda b, settings:
                    settings.set_property('recursive', b),
            'searcharchives':
                lambda b, settings:
                    settings.set_property('searcharchives', b),
            'uniquelines':
                lambda b, settings:
                    settings.set_property('uniquelines', b),
            'verbose':
                lambda b, settings:
                    settings.set_property('verbose', b),
            'version':
                lambda b, settings:
                    settings.set_property('printversion', b)
        }
        self.longarg_dict = {}

    def settings_from_file(self, filepath, settings):
        assert os.path.exists(filepath), 'Settings file not found: %s' % filepath
        with open(filepath) as f:
            jsonstr = f.read()
            # print "jsonstr: '%s'" % jsonstr
            self.settings_from_json(jsonstr, settings)

    def settings_from_json(self, jsonstr, settings):
        json_dict = json.loads(jsonstr)
        for arg in json_dict:
            if arg in self.arg_action_dict:
                self.arg_action_dict[arg](json_dict[arg], settings)
            elif arg in self.bool_flag_action_dict:
                self.bool_flag_action_dict[arg](json_dict[arg], settings)
            elif arg == 'startpath':
                settings.startpath = json_dict[arg]
            else:
                raise SearchException('Invalid option: {0}'.format(arg))

    def set_options_from_xml(self):
        searchoptionsdom = minidom.parse(SEARCHOPTIONSPATH)
        searchoptionnodes = searchoptionsdom.getElementsByTagName('searchoption')
        for searchoptionnode in searchoptionnodes:
            longarg = searchoptionnode.getAttribute('long')
            shortarg = searchoptionnode.getAttribute('short')
            desc = get_text(searchoptionnode.childNodes).strip()
            # func = None
            if longarg in self.arg_action_dict:
                func = self.arg_action_dict[longarg]
            elif longarg in self.bool_flag_action_dict:
                func = self.bool_flag_action_dict[longarg]
            else:
                raise SearchException('Unknown search option: %s' % longarg)
            self.options.append(SearchOption(shortarg, longarg, desc, func))
            self.longarg_dict[longarg] = longarg
            if shortarg:
                self.longarg_dict[shortarg] = longarg

    def search_settings_from_args(self, args):
        """Returns a SearchSettings instance for a given list of args"""
        settings = SearchSettings()
        # default printresults to True since running from command line
        settings.printresults = True
        argdeque = deque(args)
        while argdeque:
            arg = argdeque.popleft()
            if arg.startswith('-'):
                while arg and arg.startswith('-'):
                    arg = arg[1:]
                if arg in self.longarg_dict:
                    longarg = self.longarg_dict[arg]
                    if longarg in self.arg_action_dict:
                        if argdeque:
                            argval = argdeque.popleft()
                            self.arg_action_dict[longarg](argval, settings)
                        else:
                            raise SearchException('Missing value for option {0}'.
                                            format(arg))
                    elif longarg in self.bool_flag_action_dict:
                        self.bool_flag_action_dict[longarg](True, settings)
                        if longarg in ('help', 'version'):
                            return settings
                    else:
                        raise SearchException('Invalid option: {0}'.format(arg))
                else:
                    raise SearchException('Invalid option: {0}'.format(arg))
            else:
                settings.startpath = arg
        return settings

    def usage(self):
        print(self.get_usage_string())
        sys.exit(1)

    def get_usage_string(self):
        sio = StringIO()
        sio.write('Usage:\n')
        sio.write(' pysearch.py [options] -s <searchpattern> <startpath>\n\nOptions:\n')
        opt_strings = []
        opt_descs = []
        longest = 0
        for opt in sorted(self.options, key=lambda o: o.sortarg):
            opt_string = ''
            if opt.shortarg:
                opt_string += '-{0},'.format(opt.shortarg)
            opt_string += '--{0}'.format(opt.longarg)
            if len(opt_string) > longest:
                longest = len(opt_string)
            opt_strings.append(opt_string)
            opt_descs.append(opt.desc)
        format_string = ' %%-%ds  %%s\n' % longest
        for i, opt_string in enumerate(opt_strings):
            sio.write(format_string % (opt_string, opt_descs[i]))
        usage = sio.getvalue()
        sio.close()
        return usage

File: ruby/rbsearch/searchoptions.rb
################################################################################
#
# searchoptions.rb
#
# class SearchOptions: defines the available command-line options and
#                      corresponding utility methods
#
################################################################################

require 'json'
require 'rexml/document'
include REXML
require_relative 'config.rb'
require_relative 'searchoption.rb'
require_relative 'searchsettings.rb'

class SearchOptions

  def initialize
    @options = []
    @arg_action_dict = {}
    @bool_flag_action_dict = {}
    @longarg_dict = {}
    set_actions
    set_options_from_xml
    @options.sort! { |a, b| a.sortarg <=> b.sortarg }
  end

  def set_actions
    @arg_action_dict = {
      :'in-archiveext' => ->(x, settings){ settings.add_exts(x,
        settings.in_archiveextensions) },
      :'in-archivefilepattern' => ->(x, settings){ settings.add_patterns(x,
        settings.in_archivefilepatterns) },
      :'in-dirpattern' => ->(x, settings){ settings.add_patterns(x,
        settings.in_dirpatterns) },
      :'in-ext' => ->(x, settings){ settings.add_exts(x,
        settings.in_extensions) },
      :'in-filetype' => ->(x, settings){ settings.add_filetypes(x,
        settings.in_filetypes) },
      :'in-filepattern' => ->(x, settings){ settings.add_patterns(x,
        settings.in_filepatterns) },
      :'in-linesafterpattern' => ->(x, settings){ settings.add_patterns(x,
        settings.in_linesafterpatterns) },
      :'in-linesbeforepattern' => ->(x, settings){ settings.add_patterns(x,
        settings.in_linesbeforepatterns) },
      :'linesafter' => ->(x, settings){ settings.linesafter = x.to_i },
      :'linesaftertopattern' => ->(x, settings){ settings.add_patterns(x,
        settings.linesaftertopatterns) },
      :'linesafteruntilpattern' => ->(x, settings){ settings.add_patterns(x,
        settings.linesafteruntilpatterns) },
      :'linesbefore' => ->(x, settings){ settings.linesbefore = x.to_i },
      :'maxlinelength' => ->(x, settings){ settings.maxlinelength = x.to_i },
      :'out-archiveext' => ->(x, settings){ settings.add_exts(x,
        settings.out_archiveextensions) },
      :'out-archivefilepattern' => ->(x, settings){ settings.add_patterns(x,
        settings.out_archivefilepatterns) },
      :'out-dirpattern' => ->(x, settings){ settings.add_patterns(x,
        settings.out_dirpatterns) },
      :'out-ext' => ->(x, settings){ settings.add_exts(x,
        settings.out_extensions) },
      :'out-filepattern' => ->(x, settings){ settings.add_patterns(x,
        settings.out_filepatterns) },
      :'out-filetype' => ->(x, settings){ settings.add_filetypes(x,
        settings.out_filetypes) },
      :'out-linesafterpattern' => ->(x, settings){ settings.add_patterns(x,
        settings.out_linesafterpatterns) },
      :'out-linesbeforepattern' => ->(x, settings){ settings.add_patterns(x,
        settings.out_linesbeforepatterns) },
      :'search' => ->(x, settings){ settings.add_patterns(x,
        settings.searchpatterns) },
      :'settings-file' => ->(x, settings){ settings_from_file(x, settings) }
    }
    @bool_flag_action_dict = {
      :'allmatches' => ->(b, settings){ settings.firstmatch = not(b) },
      :'archivesonly' => ->(b, settings){ settings.set_archivesonly(b) },
      :'caseinsensitive' => ->(b, settings){ settings.casesensitive = not(b) },
      :'casesensitive' => ->(b, settings){ settings.casesensitive = b },
      :'debug' => ->(b, settings){ settings.set_debug(b) },
      :'excludehidden' => ->(b, settings){ settings.excludehidden = b },
      :'firstmatch' => ->(b, settings){ settings.firstmatch = b },
      :'help' => ->(b, settings){ settings.printusage = b },
      :'includehidden' => ->(b, settings){ settings.excludehidden = not(b) },
      :'listdirs' => ->(b, settings){ settings.listdirs = b },
      :'listfiles' => ->(b, settings){ settings.listfiles = b },
      :'listlines' => ->(b, settings){ settings.listlines = b },
      :'multilinesearch' => ->(b, settings){ settings.multilinesearch = b },
      :'noprintmatches' => ->(b, settings){ settings.printresults = not(b) },
      :'norecursive' => ->(b, settings){ settings.recursive = not(b) },
      :'nosearcharchives' => ->(b, settings){ settings.searcharchives = not(b) },
      :'printmatches' => ->(b, settings){ settings.printresults = b },
      :'recursive' => ->(b, settings){ settings.recursive = b },
      :'searcharchives' => ->(b, settings){ settings.searcharchives = b },
      :'uniquelines' => ->(b, settings){ settings.uniquelines = b },
      :'verbose' => ->(b, settings){ settings.verbose = b },
      :'version' => ->(b, settings){ settings.printversion = b }
    }
    @longarg_dict = {}
  end

  def settings_from_file(filepath, settings)
    # TODO: verify file exists
    json = File.read(filepath)
    settings_from_json(json, settings)
  end

  def settings_from_json(json, settings)
    json_hash = JSON.parse(json)
    json_hash.each do |arg, elem|
      argsym = arg.to_sym
      if @arg_action_dict.has_key?(argsym)
        @arg_action_dict[argsym].call(json_hash[arg], settings)
      elsif @bool_flag_action_dict.has_key?(argsym)
        @bool_flag_action_dict[argsym].call(json_hash[arg], settings)
        if ['h', 'help', 'V', 'version'].include?(arg)
          return
        end
      elsif arg == 'startpath'
        settings.startpath = json_hash[arg]
      else
        raise ArgumentError, "Invalid option: #{arg}"
      end
    end
  end

  def set_options_from_xml
    doc = Document.new(File.new(File.expand_path(SEARCHOPTIONSPATH)))
    doc.elements.each('searchoptions/searchoption') { |searchoption|
      long = searchoption.attributes['long']
      longsym = long.to_sym
      short = searchoption.attributes['short']
      desc = searchoption.text.strip
      func = nil
      if @arg_action_dict.has_key?(longsym)
        func = @arg_action_dict[longsym]
      elsif @bool_flag_action_dict.has_key?(longsym)
        func = @bool_flag_action_dict[longsym]
      else
        raise ArgumentError, "Unknown search option: #{long}"
      end
      @options.push(SearchOption.new(short, long, desc, func))
      @longarg_dict[long] = longsym
      if short
        @longarg_dict[short] = longsym
      end
    }
  end

  def search_settings_from_args(args)
    settings = SearchSettings.new()
    settings.printresults = true
    while args.count > 0
      arg = args.shift
      if arg.start_with?('-')
        while arg and arg.start_with?('-')
          arg = arg[1..arg.length]
        end
        longarg = @longarg_dict[arg]
        if @arg_action_dict.has_key?(longarg)
          if args.count > 0
            argval = args.shift
            @arg_action_dict[longarg].call(argval, settings)
          else
            raise ArgumentError, "Missing value for option #{arg}"
          end
        elsif @bool_flag_action_dict.has_key?(longarg)
          # pushts "arg in @flag_dict\n"
          @bool_flag_action_dict[longarg].call(true, settings)
          if ['help', 'version'].include?(longarg)
            return settings
          end
        else
          raise ArgumentError, "Invalid option: #{arg}"
        end
      else
        settings.startpath = arg
      end
    end
    settings
  end

  def usage
    puts "#{get_usage_string}\n"
    abort
  end

  def get_usage_string
    usage = ""
    usage += ("Usage:\n")
    usage += (" rbsearch.rb [options] -s <searchpattern> <startpath>\n\nOptions:\n")
    opt_strings = []
    opt_descs = []
    longest = 0
    @options.each do |opt|
      opt_string = ''
      if not opt.shortarg.empty?
        opt_string += "-#{opt.shortarg},"
      end
      opt_string += "--#{opt.longarg}"
      if opt_string.length > longest
        longest = opt_string.length
      end
      opt_strings.push(opt_string)
      opt_descs.push(opt.desc)
    end
    format_string = " %-#{longest}s  %s\n"
    i = 0
    while i < opt_strings.count
      usage += sprintf(format_string, opt_strings[i], opt_descs[i])
      i += 1
    end
    usage
  end

end
